src/lib.rs

```diff
@@ -288,18 +381,28 @@ impl Cascade {
     /// # Arguments
     /// `entry` - The slice of bytes to test for
     pub fn has(&self, entry: &[u8]) -> bool {
-        let result = self.has_internal(entry);
+        let result = match self.filter.hash_algorithm {
+            HashAlgorithm::MurmurHash3 => {
+                self.has_internal(&mut MurmurHash3IndexGenerator::new(entry))
+            }
+            HashAlgorithm::Sha256l32 => {
+                self.has_internal(&mut SHA256l32IndexGenerator::new(&self.salt, entry))
+            }
+            HashAlgorithm::Sha256 => {
+                self.has_internal(&mut SHA256CtrIndexGenerator::new(&self.salt, entry))
+            }
+        };
         if self.inverted {
             return !result;
         }
         result
     }
 
-    fn has_internal(&self, entry: &[u8]) -> bool {
-        if self.filter.has(entry, self.salt.as_ref()) {
+    fn has_internal<T: CascadeIndexGenerator>(&self, generator: &mut T) -> bool {
+        if self.filter.has(generator) {
             match self.child_layer {
                 Some(ref child) => {
-                    let child_value = !child.has_internal(entry);
+                    let child_value = !child.has_internal(generator);
```

Comment:
What do you think about putting the call to `generator.next_layer();` here, since `Cascade` knows that there are multiple `Bloom`s, but `Bloom` doesn't necessarily have any idea about that structure.

Reply:
Much clearer this way, thanks!